sub-packages/bionemo-testing/src/bionemo/testing/testing_callbacks.py

```diff
@@ -14,171 +14,239 @@
+        output: Any,
+    ) -> None:
+        """Get consumed samples as metadata."""
+        if step.trainer.training:
+            self.data.append(recursive_detach(output))
+
+
+class ValidOutputCallback(BaseInterruptedVsContinuousCallback):
+    """Collect validation output samples for comparison."""
+
+    def on_megatron_microbatch_end(
+        self,
+        step: MegatronStep,
+        batch: DataT,
+        forward_callback: "MegatronLossReduction",
+        output: Any,
+    ) -> None:
+        """Get consumed samples as metadata."""
+        if step.trainer.validating:
+            self.data.append(recursive_detach(output))
+
+
+class TrainLossCallback(BaseInterruptedVsContinuousCallback):
+    """Collect training loss samples for comparison."""
+
+    def on_megatron_step_end(
+        self,
+        step: MegatronStep,
+        microbatch_outputs: List[Any],
+        reduced: Optional[Union[torch.Tensor, Dict[str, torch.Tensor]]] = None,
+    ) -> None:
+        """Get consumed samples as metadata."""
+        if step.trainer.training:
+            self.data.append(recursive_detach(reduced))
+
+
+class ValidLossCallback(BaseInterruptedVsContinuousCallback):
+    """Collect training loss samples for comparison."""
+
+    def on_megatron_step_end(
+        self,
+        step: MegatronStep,
+        microbatch_outputs: List[Any],
+        reduced: Optional[Union[torch.Tensor, Dict[str, torch.Tensor]]] = None,
+    ) -> None:
+        """Get consumed samples as metadata."""
+        if step.trainer.training:
+            self.data.append(recursive_detach(reduced))
+
+
+class OptimizerStateCallback(BaseInterruptedVsContinuousCallback):
+    """Stop-and-go callback to check optimizer states before pausing and after resuming training."""
+
+    def on_megatron_step_start(self, step: MegatronStep) -> MegatronStep:
+        """Get optimizer states as metadata."""
+        if step.trainer.training:
+            self.data.append(
+                recursive_detach(
+                    [
+                        optimizer.mcore_optimizer.optimizer.state_dict()["state"]
+                        for optimizer in step.trainer.optimizers
+                    ]
+                )
+            )
+        return step
+
+
+class AbstractStopAndGoCallback(ABC, BaseInterruptedVsContinuousCallback):
+    """Abstract base class for stop-and-go callback to compare metadata before pausing and after resuming training.
+
+    This base class provides utility methods to help streamline stop and go comparison.
+
+    Provided methods:
+        - __init__: initializes the callback with the given mode.
+        - get_metadata: abstract method that should be overridden to get metadata from the trainer and pl_module.
+
+    Default behaviors:
+        - in stop mode, metadata is gotten and compared on_validation_epoch_end.
+        - in go mode, metadata is gotten and saved on_train_epoch_start.
+
+    Override these behaviors if necessary.
+    """
```

Comment:
I was able to refactor all the remaining callbacks except `TrainValInitConsumedSamplesStopAndGoCallback` to essentially compare states from the interrupted against the continuous training versions.

If we didn't want this callback, or wanted to refactor it, we could probably remove this class